File: odroid/Share/src/bert/share/control/Appendage.java
/**
 * Copyright 2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.share.control;

/**
 * These are the canonical names for appendages on various limbs.
 * These are body parts where we need locations.
 */
public enum Appendage
{
	LEFT_EAR,
	LEFT_EYE,
	LEFT_FINGER,
	LEFT_HEEL,
	LEFT_TOE,
	NOSE,
	RIGHT_EAR,
	RIGHT_EYE,
	RIGHT_FINGER,
	RIGHT_HEEL,
	RIGHT_TOE,
	UNKNOWN
	;

	/**
	 * Convert the Limb enumeration to text that can be pronounced.
	 * @param limb the enumeration
	 * @return user-recognizable text
	 */
	public static String toText(Appendage limb) {
		String text = "";
		switch( limb ) {
			case LEFT_EAR: text = "left ear"; break;
			case LEFT_EYE: text = "left eye"; break;
			case LEFT_FINGER: text = "left finger"; break;
			case LEFT_HEEL: text = "left heel"; break;
			case LEFT_TOE: text = "left toe"; break;
			case NOSE: text = "nose"; break;
			case RIGHT_EAR: text = "right ear"; break;
			case RIGHT_EYE: text = "right eye"; break;
			case RIGHT_FINGER: text = "right finger"; break;
			case RIGHT_HEEL: text = "right heel"; break;
			case RIGHT_TOE: text = "right toe"; break;
			case UNKNOWN: text = "unknown"; break;
		}
		return text;
	}
	/**
	 * @return  a comma-separated list of all block states in a single String.
	 */
	public static String names()
	{
		StringBuffer names = new StringBuffer();
		for (Appendage type : Appendage.values())
		{
			names.append(type.name()+", ");
		}
		return names.substring(0, names.length()-2);
	}
}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/model/Appendage.kt
/**
 * Copyright 2022-2025. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.common.model

import chuckcoughlin.bert.common.util.TextUtility

/**
 * These are the canonical names for appendages or end effectors
 * on various limbs. These are body parts where we need locations.
 * At this point there are no actions associated with them.
 */
enum class Appendage {
    LEFT_EAR, LEFT_EYE, LEFT_FINGER, LEFT_HEEL, LEFT_TOE, NOSE,
    RIGHT_EAR, RIGHT_EYE, RIGHT_FINGER, RIGHT_HEEL, RIGHT_TOE, NONE;

    companion object {
        /**
         * Convert the Limb enumeration to text that can be pronounced.
         * @param limb the enumeration
         * @return user-recognizable text
         */
        fun toText(limb: Appendage): String {
            var text = ""
            when (limb) {
                LEFT_EAR -> text = "left ear"
                LEFT_EYE -> text = "left eye"
                LEFT_FINGER -> text = "left finger"
                LEFT_HEEL -> text = "left heel"
                LEFT_TOE -> text = "left toe"
                NOSE -> text = "nose"
                RIGHT_EAR -> text = "right ear"
                RIGHT_EYE -> text = "right eye"
                RIGHT_FINGER -> text = "right finger"
                RIGHT_HEEL -> text = "right heel"
                RIGHT_TOE -> text = "right toe"
                NONE -> text = "unknown"
            }
            return text
        }

        /**
         * @return  a comma-separated list of all block states in a single String.
         */
        fun names(): String {
            val names = StringBuffer()
            for (type in values()) {
                names.append(type.name + ", ")
            }
            return names.substring(0, names.length - 2)
        }
        /** @return  a comma-separated list of common names for the extremities.
         */
        fun nameList(): String {
            val list = mutableListOf<String>()
            for (ext in Appendage.values()) {
                if( ext==Appendage.NONE) continue
                list.add(Appendage.toText(ext))
            }
            return TextUtility.createTextForSpeakingFromList(list)
        }
        /**
         * The enumeration function valueOf appears to always throw an exception.
         * This is the replacement. It is case-insensitive,
         */
        fun fromString(arg: String): Appendage {
            for (type in Appendage.values()) {
                if (type.name.equals(arg, true)) return type
            }
            return Appendage.NONE
        }
    }
}